sw/device/tests/sim_dv/uart_tx_rx_test.c

```diff
@@ -48,7 +48,17 @@ typedef enum uart_direction {
   kUartReceive,
 } uart_direction_t;
 
-// TODO: Make these datasets random.
+// These datasets will be overwritten by DV test.
+
+// There are multiple uart instances in the chip. DV test will randomly select
+// one and overwrite base address and interrupt IDs below.
+static volatile uint32_t enabled_uart_base_addr    = TOP_EARLGREY_UART0_BASE_ADDR;
+static volatile uint32_t uart_irq_tx_watermartk_id = kTopEarlgreyPlicIrqIdUart1TxWatermark;
+static volatile uint32_t uart_irq_rx_watermartk_id = kTopEarlgreyPlicIrqIdUart1RxWatermark;
+static volatile uint32_t uart_irq_tx_empty_id      = kTopEarlgreyPlicIrqIdUart1TxEmpty;
+static volatile uint32_t uart_irq_rx_overflow_id   = kTopEarlgreyPlicIrqIdUart1RxOverflow;
+
+static const uint8_t enabled_uart_base_addr_array[4] = {0x00, 0x00, 0x01, 0x40};
```

Comment:
The variables I added here are meant to be overridden. First we use `/usr/bin/readelf` to know where the memory location are for these variables, then backdoor override the memory content.
I did a few experiments.
1. use `static const uint32_t uart_irq_tx_watermartk_id = kTopEarlgreyPlicIrqIdUart1TxWatermark;`, but the variable `uart_irq_tx_watermartk_id` doesn't appear in the output file from `/usr/bin/readelf`. I guess it's somehow optimized and no need a dedicated memory location for it. Please correct me if you have a way to find the memory location of this variable.
2. switch `const` to `volatile`, then this variable appears in the output file from `/usr/bin/readelf` as below but it is stored at main_ram. When I override that address (I confirmed the memory is updated correctly), the SW doesn't use the value I override. Is it because the way I use the variable isn't right? (I use it as a value rather than a pointer to that location). Or we shouldn't backdoor modify main_ram?
>   1693: 10000000     4 OBJECT  LOCAL  DEFAULT    7 uart_irq_tx_watermartk_id

3. I make it as an array. and it's stored in flash, which works. But this approach is not pretty. Most of variables are not arrays. Prefer to find a better approach.
>   1622: 20003d28     4 OBJECT  LOCAL  DEFAULT    6 enabled_uart_base_addr_array

@sriyerg @mcy any suggestion?

Reply:
Why can't you just pick magic memory locations and have the simulation scribble on those specifically, and then do e.g.

`static uint32_t *my_magic_param = (uint32_t *) 0xmagic;`

That seems far more consistent than playing chicken with the linker.

Reply:
> Why can't you just pick magic memory locations and have the simulation scribble on those specifically, and then do e.g.
> 
> `static uint32_t *my_magic_param = (uint32_t *) 0xmagic;`
> 
> That seems far more consistent than playing chicken with the linker.

Thanks @mcy
This approach is certain about the memory locations. And we don't need `/usr/bin/readelf` to know the address. But we need to define what address is used for the random variable and then backdoor override the value in that address in the memory.
How can we assign initial value to this address? if we don't override it, the initial value should be also valid. I think the backdoor override happen after SW image is loaded to memory and before the main program is executed. So I feel we need something like this `static uint32_t &my_magic_param = 1;` to assign initial value for that location, but this doesn't work.

@sriyerg any thoughts?